### ISSUE TITLE
Streaming with tailable query is crashing once exhausted in version 4.0.*

### ISSUE THREAD

**Issue by fazilvp** [Automattic/mongoose#2922]
Below code is crashing in version 4.0.2 after all documents are exhausted. It was working in 3.8.27. Here is a [Live Demo](http://runnable.com/VT4syHtLGv8FqOdf/%5Berror%5D-mongoose-streaming-with-tailable-cursor-and-capped-collection) of the issue.

``` sh
stream = User.find().tailable().stream();
stream.on("data",function(doc) {
    console.log( "Streamed:\n%s", doc );
});
```

Below is the error.

``` sh
/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:237
      throw err;
            ^
Error
    at Error.MongoError (/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/error.js:13:17)
    at Function.MongoError.create (/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/error.js:31:11)
    at /root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/cursor.js:563:54
    at queryCallback (/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/wireprotocol/2_4_support.js:109:5)
    at Callbacks.emit (/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:84:3)
    at null.messageHandler (/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/topologies/server.js:218:23)
    at Socket.<anonymous> (/root/node_modules/mongoose/node_modules/mongodb/node_modules/mongodb-core/lib/connection/connection.js:259:22)
    at Socket.EventEmitter.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:736:14)
    at Socket.EventEmitter.emit (events.js:92:17)
```

**Comment by kelonye** [comment]
The error is not thrown if you listen to the "error" event:

```
      stream.on('error', function(error) {
        console.error(val);
      });
```

**Comment by kelonye** [comment]
You will then have to recreate the stream:

```javascript


var lastTimestamp = moment.utc().subtract(60, 'minute').valueOf();

fn();

function fn(){

  var opts = {
    timestamp: { $gt: lastTimestamp },
  }

  var stream = Collection.find(opts).tailable().stream();

  stream.on('data', function(doc){
    lastTimestamp = doc.timestamp; // date field
    console.log(data);
  });

  stream.on('error', function(val) {
    console.error(val);
    stream.emit('close');
    fn();
  });

}
```

**Comment by vkarpov15** [comment]
@kelonye is right, that's unfortunately the way to handle this right now because of changes to the mongodb driver. Mongoose >= 4 uses mongodb driver 2.0, which only retries tailable cursors up to a certain point: http://mongodb.github.io/node-mongodb-native/2.0/api/Cursor.html#setCursorOption

**Comment by maxs15** [comment]
Anything new about this issue ?
Spamming creation of streams looks like a really dirty solution :/

**Comment by vkarpov15** [comment]
```javascript
var stream = Model
    .find(query)
    .tailable(true, {
        awaitdata: false,
        numberOfRetries: Number.MAX_VALUE,
        tailableRetryInterval: 200
    })
```

Added to docs above

**Comment by maxs15** [comment]
Thanks @vkarpov15  for the quick answer !
So if I understand there is still no way to keep the stream open ? Furthermore if you reach the numberOfRetries limit the stream will stop

**Comment by vkarpov15** [comment]
You might be able to use Number.POSITIVE_INFINITY instead, which would keep the stream open forever, both are equivalent for practical purposes though, MAX_VALUE is on the order of 10^308.

**Comment by jbasiago** [comment]
I've been using tailable cursors with the techniques described above, but I find that cpu utilization is extraordinarily high when I do.  If I turn off the stream, my cpu idles at around 1% utilization.  Just turning on the streaming raises it to 9% without any activity in the system at all.  It appears to be somewhere deep in node, but I don't know how to troubleshoot it.  I'm currently using node version 4.2.3 and mongoose 4.4.8.  Any ideas?  Changing the tailableRetryInterval has no affect on the cpu utilization.  Thanks.

**Comment by markmeeus** [comment]
@jbasiago What version of Mongo are you running?  From this ticket: https://jira.mongodb.org/browse/SERVER-9580 I read that any version below 3.1.6 makes the client use a busy-waiting mechanism to wait for new data.  Which would explain your cpu usage.

**Comment by jbasiago** [comment]
Thanks for pointing that out.  I hadn't seen that ticket.  We are running 2.6.11 in production.  We're long overdue for an upgrade there.  As I said, we have been using single instance tailable cursors for some time without seeing a significantly high cpu utilization, but it seems that version 4.4.8 really exacerbates that condition.  We'll stick with 4.4.7 until we can upgrade our version of Mongo.  Thanks again.

**Comment by Carpetfizz** [comment]
Has this issue been fixed yet? I'm using the latest version of mongoose and MongoDB 3.2.5 and I'm still getting this error. This is how I'm creating my stream:

```
var processedPhotoStream = ProcessedPhoto.find().tailable(true, 
	{awaitdata: true, numberOfRetries: Number.MAX_VALUE}).stream();
```

I get the `No more documents in tailed cursor` error when there are no documents or they have all been exhausted. I wanted it to wait for new documents.

**Comment by vkarpov15** [comment]
Re: comments in https://github.com/Automattic/mongoose/issues/3109#issuecomment-274717171, should be fixed with mongoose >= 4.4.8
